### PR TITLE
Fix for #190

### PR DESCRIPTION
**Pull request by Khazrak**
I forked jgriffiths1993's DockerImageName-class and implemented it in the Commands.

Dockers remote API is not consistent when a tag is just the last part or when a tag is a full image name with registry.

Unit tests succeed but it might need more "real" tests.
Please review and request any changes.

---

This change is [Reviewable](https://reviewable.io/reviews/docker-java/docker-java/239)

### CODE REVIEW

File: src/main/java/com/github/dockerjava/core/util/DockerImageName.java [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+ * [REGISTRY/][NAMESPACE/]REPOSITORY[:TAG]
+ * 
+ * Based on Docker v1.5 source
+ * @author jgriffiths 08/04/2015
+ */
+public class DockerImageName {
+    
+    public String imageName;
+    
+    private TagNamePart tag = null;
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/core/util/DockerImageName.java]: they will be null by default

File: src/main/java/com/github/dockerjava/core/util/DockerImageName.java [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+        this.imageName = imageName;
+        // If we have a string to construct with
+        if( imageName != null && !imageName.equals("") ) {
+            // Split the name into a map and construct the image name parts
+            // This action will throw an exception if the name contains too many parts
+            Map<String, String> nameMap = splitImageName(imageName);
+            if( nameMap.containsKey("registry") ) {
+                this.registry = new RegistryNamePart(nameMap.get("registry"));
+            }
+            if( nameMap.containsKey("namespace") ) {
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/core/util/DockerImageName.java]: space after if and not before variable

Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/core/util/DockerImageName.java]: or it convention surround with spaces?

File: src/main/java/com/github/dockerjava/core/util/DockerImageName.java [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+    }
+    
+    public void makeValid() {
+        if( registry != null ) {
+            registry.makeValid();
+        }
+        if( namespace != null ) {
+            namespace.makeValid();
+        }
+         if( repository != null ) {
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/core/util/DockerImageName.java]: indent

File: src/main/java/com/github/dockerjava/core/util/DockerImageNameException.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+/**
+ *
+ * @author jgriffiths1993
+ */
+public class DockerImageNameException extends IllegalArgumentException {
+    public DockerImageNameException(String message) {
+        super(message);
+    }
+}
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/core/util/DockerImageNameException.java]: new line

File: src/main/java/com/github/dockerjava/jaxrs/TagImageCmdExec.java [hunk 5]
```diff
@@ -18,16 +19,40 @@ public TagImageCmdExec(WebTarget baseResource) {
...
 		return null;
 	}
-	
+
+	/**
+	 * Returns a String with all parts except Tag
+	 * @param imageName
+	 * @return
+	 */
+	private String getImageNameWithoutTag(DockerImageName imageName)
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/jaxrs/TagImageCmdExec.java]: Why not put it in getter for DockerImageName?

Comment (Khazrak) [inline src/main/java/com/github/dockerjava/jaxrs/TagImageCmdExec.java]: In this case I'm unsure of the Remote API, it seems to want the new name split into repo (all but the tag) and the tag.

**Comment by KostyaSha** [comment]
Oh, seems i accidentally tried to re-implement the same idea in #364 
Would you like to try rework this PR with following existed coding style?

**Comment by Khazrak** [comment]
Will take a look at it later today

**Comment by Khazrak** [comment]
I've reimplemented jgriffiths old implementation and started to replace Identifier and Repository class with DockerImageName.

2 questions: 
He put DockerImageName in core/utils but it feels more like a core/model to me, where should it be?
In the remote API and the code "repo" is used alot and I'm not sure if this means just repository or the whole registry/namespace/repository part?

Example:
Original 

`.queryParam("repo", command.getRepository()).queryParam("tag", command.getTag())`

Updated
```
.queryParam("repo", command.getDockerImageName().toStringWithoutTag())
.queryParam("tag", command.getDockerImageName().getTag().toString())
```
Where "toStringWithoutTag()" returns all the parts without tag that arn't null (registry/namespace/repository)

ps: did a new fork of the docker-java/docker-java git repo due to the old one was more work to merge then to redo all, so when the time come this will be a new PR

**Comment by KostyaSha** [comment]
imho If request model allows separate string fields, then this single methods should be kept. Additional setter/getter should provide this new helper dockerimagename object. 
Then in `*exec` if it would allow simplify some logic (parsing, extracting part of name) it can be used, and it will be again helper/util. 

> link to code: https://github.com/Khazrak/docker-java/

Please make PR.

**Comment by Khazrak** [comment]
I did a PR #535
It doesn't have any code for changes more then was required for the removal of Identifier and Repository classes. The Cmd's still has their 'repo' and 'tag' String as before

**Comment by bsideup** [comment]
Given the age, number of conflicts and the amount of breaking changes, I believe we have to close this PR.

I like the idea tho, would be cool to see it re-implemented in the future if somebody wants to work on it!

**Pull request closed without merging**